cpp/src/arrow/csv/reader_test.cc

```diff
@@ -216,5 +216,83 @@ TEST(StreamingReaderTests, NestedParallelism) {
   TestNestedParallelism(thread_pool, table_factory);
 }
 
+TEST(ReaderOptionsTests, SkipRowsAfterNames) {
```

Comment:
I can look into writing them in python. I just wrote the test in C++ because I implemented the feature in C++.
Right now I am actually not working on this because after the above discussions I was under the impression that the desired implementation would be to have skip_rows be a list of ranges of row indexes to skip. With that in mind I actually started to work on https://issues.apache.org/jira/browse/ARROW-12675 so that the parser could know what absolute line number it was parsing and could skip the line.

Reply:
IMHO a list of ranges would be overkill. `skip_rows_after_names` seems ok to me.

Reply:
Actually after looking at it a bit more it doesn't have to be moved out of the reader but I don't think it can use SkipRows. SkipRows is very simple in its implementation it doesn't actually skip rows but lines in the file. I make the distinction here because a row can contain values with new lines. If any of the rows contain a quoted or escaped new line skip rows will consider that two lines and not one.

I was thinking it might be better to add add a FirstN method to Chunker to be able to get the Nth occurrence of the line endings. I was thinking this could be integrated into the BlockReader implementations to be able to skip over rows even beyond the first block. This could also solve ARROW-8527.

Reply:
The fact that SkipRows is a bit dumb is a separate concern. We don't seem to have had any user complaints about it, so no need to complicate it for now, IMHO.

Reply:
My guess is that the reason it is simple is because of the comment that it is intended to skip corrupt rows so for that it probably has to be a bit brute force.

While adding a more csv aware skip does add some more complexity that parsing knowledge is already contained in the BoundryFinder implementations, so it already exists. Also, as a user specifying the number of rows to skip I would expect that csv rows would be skipped and not file lines.

If it sways your opinion any, I did get the implementation working that uses the BlockReaders, Chunker and BoundryFinders to skip over the lines and the parser and everything downstream are unaware. Also, it can skip lines beyond a single block, satisfying ARROW-8527.

Reply:
Well, if it's not too much code and you want to do it, then why not :-)

Reply:
I updated the PR with a version which uses the BoundryFinder and can skip rows in more than one block